peer/hashring32/ring.go

```diff
@@ -0,0 +1,338 @@
+// an individual differs between implementations. Some implementations
+// may use the hint when choosing, others might ignore it.
+//
+// A Choose on an empty population should return an error.
+func (r *ring) Choose(ctx context.Context, req *transport.Request) peer.Identifier {
+	return r.choose(req.ShardKey)
+}
+
+func (r *ring) choose(shardKey string) peer.Identifier {
+	r.m.RLock()
+	defer r.m.RUnlock()
+
+	if len(r.servers) == 0 {
+		return nil
+	}
+
+	var ix int
+	if shardKey == "" {
+		// Random index to get hash value
+		ix = rand.Intn(len(r.hashRing))
+	} else {
+		// Binary search to find hash value
+		key := r.hashFunc([]byte(shardKey))
+		ix = indexOf(r.hashRing, key)
+	}
+
+	hash := r.hashRing[ix]
+	servers := r.serversByHash[hash]
+
+	return servers[0]
+}
+
+// Add an individual to the population.
+// Returns a subscriber if the individual was absent, nil otherwise.
+func (r *ring) Add(pid peer.Identifier) peer.Subscriber {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	individual := pid.Identifier()
+
+	if _, ok := r.servers[individual]; ok {
+		return nil
+	}
+	r.addToServersMapAndHashRing(pid)
+	r.radixSortHashRing()
+	return nopSubscriber
+}
+
+// Remove removes an individual from the population.
+// Returns early if the individual is absent.
+func (r *ring) Remove(pid peer.Identifier, sub peer.Subscriber) {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	individual := pid.Identifier()
+
+	if _, ok := r.servers[individual]; !ok {
+		return
+	}
+	toBeRemovedIndexSet := make(map[int]struct{})
+	r.removeFromServersMap(pid, toBeRemovedIndexSet)
+	r.removeFromHashRing(toBeRemovedIndexSet)
+}
+
+// Include extends the ring's population with new members. Think
+// set union.
+// The implementations can mutate the input.
+func (r *ring) Include(pids []peer.Identifier) {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	r.includeNoLock(pids)
+}
+
+func (r *ring) includeNoLock(pids []peer.Identifier) {
+	for _, pid := range pids {
+		individual := pid.Identifier()
+		if _, ok := r.servers[individual]; ok {
+			continue
+		}
+		r.addToServersMapAndHashRing(pid)
+	}
+	r.radixSortHashRing()
+}
+
+// Exclude shrinks the ring's population by removing members. Think
+// set difference.
+// The implementations can mutate the input.
+func (r *ring) Exclude(pids []peer.Identifier) {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	r.excludeNoLock(pids)
+}
+
+func (r *ring) excludeNoLock(pids []peer.Identifier) {
+	toBeRemovedIndexSet := make(map[int]struct{})
+
+	for _, pid := range pids {
+		individual := pid.Identifier()
+		if _, ok := r.servers[individual]; !ok {
+			continue
+		}
+		r.removeFromServersMap(pid, toBeRemovedIndexSet)
+	}
+	r.removeFromHashRing(toBeRemovedIndexSet)
+}
+
+// Set replaces the ring's population with the new one.
+// The implementations can mutate the input.
+func (r *ring) Set(pids []peer.Identifier) {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// Create string set for population, to ensure it is addressable.
+	population := make(map[string]peer.Identifier)
```

Comment:
we can preset the capacity right?

Reply:
good catch, this should be the number of configured replicas.